src/core/router/identity.cc

```diff
@@ -729,42 +727,51 @@ Keys CreateRandomKeys() {
 
 IdentHash CreateRoutingKey(
     const IdentHash& ident) {
-  std::uint8_t buf[41];  // ident + yyyymmdd
-  memcpy(buf, (const std::uint8_t *)ident, 32);
-  time_t t = time(nullptr);
-  struct tm tm;
-  // TODO(unassigned): never use sprintf, use snprintf instead.
-#ifdef _WIN32
-  gmtime_s(&tm, &t);
-  sprintf_s(
-      reinterpret_cast<char *>((buf + 32)),
-      9,
-      "%04i%02i%02i",
-      tm.tm_year + 1900,
-      tm.tm_mon + 1,
-      tm.tm_mday);
-#else
-  gmtime_r(&t, &tm);
-  sprintf(
-      reinterpret_cast<char *>((buf + 32)),
-      "%04i%02i%02i",
-      tm.tm_year + 1900,
-      tm.tm_mon + 1,
-      tm.tm_mday);
-#endif
   IdentHash key;
-  // TODO(anonimal): this try block should be larger or handled entirely by caller
-  try {
-    kovri::core::SHA256().CalculateDigest((std::uint8_t *)key, buf, 40);
-  } catch (...) {
-    core::Exception ex;
-    ex.Dispatch(__func__);
-    // TODO(anonimal): review if we need to safely break control, ensure exception handling by callers
-    throw;
-  }
+  try
+    {
+      if (ident.IsZero())
+        {
+          throw std::invalid_argument(
+              "Identity: cannot create routing key from zero-initialized "
+              "ident");
+        }
+      // Format the routing key
+      const std::vector<std::uint8_t> formatted_buf = FormatRoutingKey(ident);
+      // Hash the formatted buffer to create the routing key
+      kovri::core::SHA256().CalculateDigest(
+          key(), formatted_buf.data(), formatted_buf.size());
+    }
+  catch (...)
+    {  // Error formatting or hashing the key
+      kovri::core::Exception ex(__func__);
+      ex.Dispatch();
+      throw;
+    }
   return key;
 }
 
+std::vector<std::uint8_t> FormatRoutingKey(const IdentHash& ident)
```

Comment:
I did not add the `boost::gregorian::date` object as a second parameter, since I thought passing it by non-const reference may lead to thread-unsafe usage. Instead, I extract the date and ident from the result in tests.

Reply:
See revert comment.